src/java/org/apache/cassandra/index/SecondaryIndexManager.java

```diff
@@ -1179,12 +1202,20 @@ private static String commaSeparated(Collection<Index> indexes)
         return indexes.stream().map(i -> i.getIndexMetadata().name).collect(Collectors.joining(","));
     }
 
-
     public Optional<Index> getBestIndexFor(RowFilter.Expression expression)
     {
         return indexes.values().stream().filter((i) -> i.supportsExpression(expression.column(), expression.operator())).findFirst();
     }
 
+    public <T extends Index> Optional<T> getBestIndexFor(RowFilter.Expression expression, Class<T> indexType)
```

Comment:
I have made this change here but have left the basic implementation the same in `QueryController` where it is used. This could, potentially, be useful when we come to a query optimiser that may be able to score indexes on their ability to service a query.